### ISSUE TITLE
[jvm-packages] Scalability question

### ISSUE THREAD

**Issue by nikolay-karpow** [dmlc/xgboost#2146]
I run XGBoost model training on Spark standalone cluster. Cluster consists from two machines 16 cores and 256 Gb RAM each. The data set for model training has about 450k elements with 70k unique features. Spark is configured to use 1 core for each executor, there are 30 executors.

I run training with different nWorkers number. Here is the training time:
nWorkers = 30 => 33 minutes
nWorkers = 15 => 30 minutes
nWorkers = 5 => 38 minutes
nWorkers = 1 => 100 minutes

So, with 15 workers the performance is slightly better, than with 30, what is counter-intuitive for me. Is this a known scaling behavior? Is there any rule of thumb of determining the optimal workers count?

## Environment info
Operating System:
CentOS 7.2

xgboost version used:
commit ef8d92fc52c674c44b824949388e72175f72e4d1 from Jan 9

Spark 2.0.2

**Comment by gjmulder** [comment]
There looks to be a serial step in your Spark processing that isn't allowing you to scale across both nodes. The extra 3 minutes may be Spark distributing your job across both nodes.

**Comment by nikolay-karpow** [comment]
It can hardly be the case, because I measure only a small part of the whole job. So, this time includes only the following:
- XGBoost.trainWithRDD call (labeled points are prepared already)
- Prediction on the training data set 
- Collect call for prediction result (here I collect just a single Double value for each data set element)

**Comment by CodingCat** [comment]
Before call trainwithrdd, the train rdd has been evaluated?

**Comment by CodingCat** [comment]
and persisted/cached?

**Comment by nikolay-karpow** [comment]
To be precise, I do the following:

1. Parse input data set into `JavaRDD<MyCustomClass> parsedRdd` and call `parsedRdd.cache()`.

2. Index features (originally I have features names, here I get ids instead) from parsedRdd and collect them as a map (so the parsing is evaluated after this step)

3. Create `JavaRDD<CustomLabeledPoint> customLabeledPoints` from parsedRDD and features map, call `customLabeledPoints.cache()`. This step includes two broadcasts and one join.

4. Create `JavaRDD<org.apache.spark.ml.feature.LabeledPoint> labeledPoints` from customLabeledPoints via very trivial mapping

5. Call `XGBoost.trainWithRDD(labeledPoints.rdd())`

6. Once the actual training is completed, 
`RDD<Vector> featuresRdd = labeledPoints.map(LabeledPoint::features).rdd();
JavaRDD<float[][]> predictions = model.predict(featuresRdd, false).toJavaRDD();`
And transform predictions to a flat `JavaRDD<Double> doublePredictions`

7. Call `List<Double> predictionsList = doublePredictions.collect()`

8. List<Double> labelesList = labeledPoints.map(org.apache.spark.ml.feature.LabeledPoint::label).collect();

9. Perform some trivial calculations with predictionsList and labelesList locally.

The measured time includes steps 3-9.

**Comment by CodingCat** [comment]
can you just measure them separately? from your description, I cannot make sure which step is the bottleneck

you may need more experiments to answer 

1. is training scalable? (to detect this, you may force the spark evaluate 3 before running 5, e.g. for the RDD in 3, rddInStep3.cache().foreach(_ => Unit) and then you can measure the pure training scalability)

and then you can further answer: 


2. is predict in 6 the bottleneck?

3. is collect in 7 the bottleneck?

4. is collect in 8 the bottleneck?

**Comment by nikolay-karpow** [comment]
OK, I've done that. Data set is cached and evaluated just before training. The measuremenet includes only XGBoost.trainWithRDD call. I've measured it on 15 and on 30 cores:
- 15 cores - 36 minutes
- 30 cores - 32 minutes
Any ideas?

**Comment by CodingCat** [comment]
I can not see the reason from these two numbers...in my experiments with various datasets, I didn't see any scalability bottleneck like this

Would you post your log? and @tqchen , do you have any idea on the scalability of XGBoost in such high dimensional vectors (70k features)

**Comment by nikolay-karpow** [comment]
I'm afraid, I'm not allowed to share the log...

**Comment by CodingCat** [comment]
did you validate the functionality of the produced model? does it work as expected?

**Comment by nikolay-karpow** [comment]
I haven't done that for this particular data set, but I've trained about 20 of others (they are smaller) with validation. So, the whole pipeline works fine, I suppose.

**Comment by tqchen** [comment]
Normally there is indeed a limitation when you have more features(networking) and less computing. Assigning more CPU per worker might help in here, since there is no networking within each worker

**Comment by tqchen** [comment]
However, this does not mean xgboost does not scale (with biggger data). It simply means there is a limitation on the number of workers we can scale to for a fixed dataset

**Comment by nikolay-karpow** [comment]
@tqchen Thanks for explanation, it makes sense. Do you have any idea of how to get optimal workers count based on data set size and features count?

**Comment by fatihtekin** [comment]
Hi, Unfortunately, I am having exactly the same issue and did as you suggested(increasing core per worker) but isn't there any other optimization like sending the results as batches over the network etc.
